iree/vm/module.h

```diff
@@ -112,25 +137,101 @@ typedef struct {
 // VM functions and accessing this state.
 typedef struct iree_vm_module_state iree_vm_module_state_t;
 
-// Function call arguments.
+// Function call data.
+//
+// Arguments and results are encoded following a standard format shared across
+// all module types. This allows implementations that have different storage
+// types (such as physical machine registers vs. virtual registers) to use the
+// same cross-module calling convention.
+//
+// Callees can assume that callers have properly allocated and setup the
+// argument and result buffers and need not verify them. This works only because
+// the calling convention format is directly queried from the callee module.
```

Comment:
Yeah, exposing multiple calling conventions seems messy. Brainstorming some more to see if there are design flaws here...

One case: An Android app has runtime code pinned to a specific version and some newer module coming over the network (my_super_smart_model_v4 totally needs a module for something). The app could ask the server for a compatible version of the module in that case, and the server could have multiple versions (trimmed for size to respect limited network bandwidth) available. If the newer module actually requires the new calling convention, then the app should handle that failures (by prompting the user to update it, etc.).

Another case: A device (Android phone, embedded system, etc.) has an old module stored in read-only-memory somewhere in its firmware. User-space code has a newer runtime library and wants to use newer features (new calling convention has X% less overhead or includes direct support for some critical type like bfloat16, etc.). In order to use the old module and the new features, the application would need to support both calling conventions?

Reply:
Similar to how an app may want to fetch a precompiled/signed AOT module for the appropriate platform they could do the same along this aspect.

Using older modules with newer runtimes (backwards compatibility of the module format) is much easier to support - the module would say it has a particular version and then the runtime can choose a different code path for v0 vs. v1.

I think on the whole I don't see this changing much in a way that would break compatibility and require a version bump. It's similar to protos/flatbuffers/etc; adding new things is easy (supporting new types) as that won't break backwards compat, but changing the behavior of existing things is hard (requires version bump and common understanding).